crates/common/src/input.rs

```diff
@@ -118,10 +78,22 @@ impl InputIngot {
     }
 }
 
-#[salsa::input]
+#[salsa::interned]
+pub struct FilePath<'db> {
+    path: Utf8PathBuf,
+}
+
+impl<'db> FilePath<'db> {
+    pub fn from(db: &'db dyn InputDb, path: impl Into<Utf8PathBuf>) -> Self {
+        FilePath::new(db, path.into())
+    }
+}
+
+#[salsa::input(constructor = __new_impl)]
 pub struct InputFile {
     /// A path to the file from the ingot root directory.
     #[return_ref]
+    #[set(__set_path_impl)]
```

Comment:
If `InputFile`s are loosely tied to `InputIngot`s, we need to discourage people from modifying `InputFile.path` directly.

This could be handled by some `file_rename` method on `InputIngot`.  It could also be real nice to replace `InputIngot::files` with a map or patricia trie to get nice cheap getters.  If we're already going funnel `InputFile::path` modification through an `InputIngot::file_rename` method, maintaining that index would be better justified.